### PR TITLE
Store `GetFullHtmlFieldName()` and `CreateSanitizedId()` values

### PR DESCRIPTION
**Pull request by dougbu**
- #3918
- don't repeat allocations for identical calls; helps w/ e.g. label / input / validation clusters
- add `NameAndIdProvider`; it stores `PreviousNameAndId` in `HttpContext.Items`
 - `PreviousNameAndId` allocated only when `string`s are allocated e.g. "id"s were sanitized

### CODE REVIEW

**Comment by dougbu** [comment]
FYI this fix is for `1.1.0-preview1`.

**Comment by dougbu** [comment]
#### Allocation changes
All numbers below are for release builds and collected bytes. Reran some scenarios a few times to eliminate noise i.e. to get results capturing the same number of requests. (Sometimes see an extra request or two captured in snapshot comparisons.) Unfortunately, noise levels remain high.

Expect reductions in the number of `string`s, `StringBuilder`s and backing `char[]`s. Other allocation changes are likely unrelated to anything in this PR.
##### BasicViews
Exercises [HtmlHelpers.cshtml](https://github.com/aspnet/Performance/blob/dev/testapp/BasicViews/Views/Home/HtmlHelpers.cshtml) view using [`.\loadtest.ps1`](https://github.com/aspnet/Performance/blob/dev/testapp/BasicViews/loadtest.ps1) `-rps 20 -variation htmlhelpers`. View uses lambda-expression-based HTML helpers in a simple "edit" form with three properties and no validation elements.
###### Savings
Snapshots show a large reduction from `dev` to this branch. But that's due to an unrelated decrease in the number of large `byte[]`s allocated.

Realistically, allocations should be identical between runs since the view and its model are very simple.
##### BigViews
Exercises [Index.cshtml](https://github.com/aspnet/Performance/blob/dev/testapp/BigViews/Views/Home/Index.cshtml) view using [`.\loadtest.ps1`](https://github.com/aspnet/Performance/blob/dev/testapp/BigViews/loadtest.ps1). View uses a partial for 10 elements of a simple collection in a simple "index" page. Partial displays five properties of each POCO element and primarily uses `Html.LabelFor()`, `Html.IdFor()`, and `Model.PropertyName`. Uses a very simple disabled `Html.ListBoxFor()` for one of the five properties (similar to default display of a `bool?`).
###### Savings
Again see a large unrelated change messing up overall comparisons. In this case, `int[]` and `byte[]` allocations go _up_ significantly for unrelated reasons.

Allocations of `string`s, `StringBuilder`s, `char[]`s, and (in the new code) `PreviousNameAndId`s go from 68.6 MB (12.1% of total allocations in that run) to 39.4 MB (7.38% of that run's allocations).
##### MusicStoreViews
Exercises [AddressAndPayment.cshtml](https://github.com/aspnet/Performance/blob/dev/testapp/MusicStoreViews/Views/Home/AddressAndPayment.cshtml) view using [`loadtest.ps1`](https://github.com/aspnet/Performance/blob/dev/testapp/MusicStoreViews/loadtest.ps1). View uses lambda-expression-based templating HTML helpers (with built-in templates) in a simple "create" form with ten fields, nine bound to a two-level POCO model.
###### Savings
Looks like runs were more consistent here. Changes save 3.9 MB or 1.08% of 363 MB baseline when using run totals.

Drilling down, allocations of `string`s, `StringBuilder`s, `char[]`s, and (in the new code) `PreviousNameAndId`s go from 37.2 MB (10.23% of total allocations in that run) to 33.2 MB (9.23% of that run's allocations).
###### Before
musicstoreviews baseline a 0
###### After
musicstoreviews cache names ids 3918 3d 0

**Comment by dougbu** [comment]
#### Requests per second changes
The numbers here look a little better in the after runs though nothing is conclusive.

Ran the listed Node.js `loadtest` commands on a Linux client. Ran builds and `dotnet {dll}` commands on a Windows server. Both machines are in Azure and neither was doing much else during the runs. In all cases, interleaved before and after runs and threw away the first `loadtest` results for each run. Runs last about 30 seconds.
##### BasicViews
This scenario is not at all CPU- or network-bound on the Azure hardware I was using. Expect the bottleneck is LocalDb but haven't investigated much.
##### BigViews
`loadtest -k http://10.0.0.7:5000/Home/Index -c 16 -n 25000`

  | run 1 | run 2 | run 3 | average | standard deviation
---- | ---- | ---- | ---- | ---- | ----
dev | 705 | 712 | 710 | 709.0 | 2.94
dougbu/cache.names.ids.3918 | 714 | 714 | 709 | 712.3 | 2.36
##### MusicStoreViews
`loadtest -k http://10.0.0.7:5000/Home/AddressAndPayment -c 16 -n 100000`

  | run 1 | run 2 | run 3 | average | standard deviation
---- | ---- | ---- | ---- | ---- | ----
dev | 3235 | 3125 | 3223 | 3194.3 | 49.27
dougbu/cache.names.ids.3918 | 3186 | 3244 | 3213 | 3214.3 | 23.70

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs [hunk 1]
```diff
@@ -29,12 +29,12 @@ public class HtmlHelper<TModel> : HtmlHelper, IHtmlHelper<TModel>
...
-                metadataProvider,
-                bufferScope,
-                htmlEncoder,
-                urlEncoder)
+                  htmlGenerator,
+                  viewEngine,
+                  metadataProvider,
+                  bufferScope,
+                  htmlEncoder,
+                  urlEncoder)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: Not sure why this changed. It looks more like what VS does but meh. I'll revert if you prefer.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/TemplateInfo.cs [hunk 2]
```diff
@@ -65,26 +65,31 @@ public bool AddVisited(object value)
             return _visitedObjects.Add(value);
         }
 
+        /// <summary>
+        /// Returns the full HTML element name for the specified <paramref name="partialFieldName"/>.
+        /// </summary>
+        /// <param name="partialFieldName">Expression name, relative to the current model.</param>
+        /// <returns>Fully-qualified expression name for <paramref name="partialFieldName"/>.</returns>
         public string GetFullHtmlFieldName(string partialFieldName)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/TemplateInfo.cs]: Changes here are just about added doc comments and consistency w/ the `NameAndIdProvider` version.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOptions.cs [hunk 3]
```diff
@@ -23,21 +26,36 @@ public class HtmlHelperOptions
...
+        {
+            get
+            {
+                return _idAttributeDotReplacement;
+            }
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOptions.cs]: `throw`ing when we call `GetFullHtmlFieldName()` was too late. Much more common for users to change `HtmlHelperOptions` than for them to call that method directly.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Rendering/TagBuilder.cs [hunk 4]
```diff
@@ -93,7 +93,9 @@ public void AddCssClass(string value)
         /// <summary>
         /// Returns a valid HTML 4.01 "id" attribute value for an element with the given <paramref name="name"/>.
         /// </summary>
-        /// <param name="name">The original element name.</param>
+        /// <param name="name">
+        /// The fully-qualified expression name, ignoring the current model. Also the original HTML element name.
+        /// </param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Rendering/TagBuilder.cs]: Changes here are just about consistency w/ the `NameAndIdProvider` versions.

**Comment by dougbu** [comment]
Ping @rynowak

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/NameAndIdProvider.cs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+using System;
+using Microsoft.AspNetCore.Mvc.Rendering;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    /// <summary>
+    /// Provides cached values for "name" and "id" HTML attributes.
+    /// </summary>
+    public static class NameAndIdProvider
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/NameAndIdProvider.cs]: Can we make this `.Internal`?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOptions.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOptions.cs]: This seems like a good change

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/NameAndIdProvider.cs (hunk 5 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/NameAndIdProvider.cs]: Sure

**Comment by dougbu** [comment]
4e1ec39

**Pull request closed without merging**